app/src/main/cpp/ime/nlp/suggestion_list.cpp

```diff
@@ -14,24 +14,22 @@
  * limitations under the License.
  */
 
-#include "staged_suggestion_list.h"
+#include "suggestion_list.h"
 
 #include <utility>
 
 using namespace ime::nlp;
 
 SuggestionList::SuggestionList(size_t _maxSize) :
-    maxSize(_maxSize), internalSize(0), internalArray(new WeightedToken*[_maxSize]), isPrimaryTokenAutoInsert(false)
+    maxSize(_maxSize), internalSize(0), tokens(new std::shared_ptr<WeightedToken>[_maxSize]), isPrimaryTokenAutoInsert(false)
 {
     // Initialize the internal array to null pointers
     for (size_t n = 0; n < maxSize; n++) {
-        internalArray[n] = nullptr;
+        tokens[n] = nullptr;
     }
 }
 
-SuggestionList::~SuggestionList() {
-    delete[] internalArray;
-}
+SuggestionList::~SuggestionList() = default;
 
 bool SuggestionList::add(word_t word, freq_t freq) {
```

Comment:
I think you could move the `word`.
```suggestion
bool SuggestionList::add(word_t&& word, freq_t freq) {
```